### PR TITLE
Rubocop

### PR DESCRIPTION
**Pull request by tamird**
builds further on #595 

@sodabrew apologies for the chain of PRs :heart:

### CODE REVIEW

**Comment by brianmario** [comment]
Holy hell this is amazing haha. I love this but would it be a huge pain to split it up into smaller pull requests that are easier to grok? Maybe one to convert the rspec syntax, one fixing up the shell script syntax, one fixing all the warnings in ruby code from Rubocop, one fixing the C code, etc... ?

**Comment by tamird** [comment]
@brianmario totally, let's land #592 and #595 first, and then I'll break this one up some more

**Comment by sodabrew** [comment]
Ok, ready to rebase here.

**Comment by tamird** [comment]
Rebased (and green). I found a bug in extconf.rb that uncovered a bunch more C warnings, which I've fixed in #628 and then rebased this on top.

**Comment by sodabrew** [comment]
Ready for another rebase.

**Comment by tamird** [comment]
Rebased but Travis is refusing to build.

**Comment by sodabrew** [comment]
Looks like Travis backed up this afternoon. I'll check back tomorrow.

**Comment by tamird** [comment]
green except for a spurious failure

**Comment by tamird** [comment]
Looks like rbx is still grumpy about the 10000 statements tests

**Comment by sodabrew** [comment]
I think there's a more general Rubinius GC problem with calls to `rb_gc_mark`, I have a ticket open there.

This crash is a new one, it's in `rb_to_encoding`, Huh. I may need to file a new ticket too.

Thanks for updating the PR!

**Comment by tamird** [comment]
Rebased but rbx is still grumpy and now mysql 5.7 is throwing a fit?

**Comment by tamird** [comment]
Oh RBX is actually leaking a connection now. MySQL 5.7 says `Table 'performance_schema.session_variables' doesn't exist`

**Comment by sodabrew** [comment]
I did some reading about the MySQL 5.7.8 changes to performance-schema, but I didn't get it. https://dev.mysql.com/doc/refman/5.7/en/performance-schema-variable-table-migration.html

**Comment by tamird** [comment]
@sodabrew ok, seems like the mysql issue isn't caused by this PR, and I'm not able to reproduce the RBX issue locally. Let me know if there's anything I can help with.

**Comment by sodabrew** [comment]
Has merge conflicts - Presumably the be_an_instance_of change touches the same diff context.

**Comment by tamird** [comment]
Yep, rebased now!

File: ext/mysql2/extconf.rb [hunk 1]
```diff
@@ -87,9 +87,9 @@ def asplode lib
   asplode 'mysql.h'
 end
 
-%w{ errmsg.h mysqld_error.h }.each do |h|
+%w(errmsg.h mysqld_error.h).each do |h|
   header = [prefix, h].compact.join '/'
-  asplode h unless have_header h
+  asplode h unless have_header header
```
Comment (sodabrew) [inline ext/mysql2/extconf.rb]: Was this a bug?

File: ext/mysql2/extconf.rb [hunk 2]
```diff
@@ -102,9 +102,7 @@ def asplode lib
   -Wno-unused-function
   -Wno-declaration-after-statement
   -Wno-missing-field-initializers
-).select do |flag|
-  try_link('int main() {return 0;}', flag)
-end.each do |flag|
+).select { |flag| try_link('int main() {return 0;}', flag) }.each do |flag|
```
Comment (sodabrew) [inline ext/mysql2/extconf.rb]: I prefer the extra lines

File: ext/mysql2/extconf.rb (hunk 1 above)
Comment (tamird) [inline ext/mysql2/extconf.rb]: yes i believe it was

File: lib/mysql2.rb [hunk 3]
```diff
@@ -8,16 +8,16 @@
...
-             elsif File.exist?(File.expand_path('../vendor/libmysql.dll', File.dirname(__FILE__)))
-               # Use vendor/libmysql.dll if it exists, convert slashes for Win32 LoadLibrary
-               File.expand_path('../vendor/libmysql.dll', File.dirname(__FILE__)).gsub('/', '\\')
-             else
-               # This will use default / system library paths
-               'libmysql.dll'
-             end
+    # If this environment variable is set, it overrides any other paths
+    # The user is advised to use backslashes not forward slashes
+    ENV['RUBY_MYSQL2_LIBMYSQL_DLL']
```
Comment (sodabrew) [inline lib/mysql2.rb]: I don't recall the reason for the `dup`... maybe I was modifying the value before. Looks safe to remove.

File: lib/mysql2/client.rb [hunk 4]
```diff
@@ -105,9 +105,12 @@ def info
       self.class.info
     end
 
-    private
-      def self.local_offset
+    class << self
+      private
+
+      def local_offset
```
Comment (sodabrew) [inline lib/mysql2/client.rb]: I've always prefered the `def self.foo` syntax, is there a reason that RuboCop prefers the class << self block syntax?

Comment (tamird) [inline lib/mysql2/client.rb]: without it, the `private` above doesn't do anything.

File: spec/mysql2/result_spec.rb [hunk 5]
```diff
@@ -279,45 +279,45 @@
     end
 
     if 1.size == 4 # 32bit
-      unless RUBY_VERSION =~ /1.8/
-        klass = Time
+      klass = if RUBY_VERSION =~ /1.8/
```
Comment (sodabrew) [inline spec/mysql2/result_spec.rb]: This could even be a ternary (unless RuboCop hates those :p)

File: spec/mysql2/statement_spec.rb [hunk 6]
```diff
@@ -130,10 +130,10 @@
 
     it "should be able to retrieve utf8 field names correctly" do
       stmt = @client.prepare 'SELECT * FROM `テーブル`'
-      expect(stmt.fields).to eq(['整数', '文字列'])
+      expect(stmt.fields).to eq(%w(整数 文字列))
```
Comment (sodabrew) [inline spec/mysql2/statement_spec.rb]: utf8 ftw.

File: ext/mysql2/extconf.rb (hunk 2 above)
Comment (tamird) [inline ext/mysql2/extconf.rb]: done

File: support/mysql_enc_to_ruby.rb [hunk 7]
```diff
@@ -65,10 +65,10 @@
 end
 
 encodings_with_nil = encodings_with_nil.map do |encoding|
-  name = "NULL"
-
-  if !encoding.nil? && encoding[1] != "NULL"
-    name = "\"#{encoding[1]}\""
+  name = if encoding.nil? || encoding[1] == 'NULL'
```
Comment (sodabrew) [inline support/mysql_enc_to_ruby.rb]: nice improvement to readability here.

File: spec/mysql2/result_spec.rb (hunk 5 above)
Comment (tamird) [inline spec/mysql2/result_spec.rb]: I'll leave that one up to you. 6 months on a PR is long enough :)

File: lib/mysql2/client.rb (hunk 4 above)
Comment (sodabrew) [inline lib/mysql2/client.rb]: I learn something new every day!

File: spec/mysql2/result_spec.rb (hunk 5 above)
Comment (sodabrew) [inline spec/mysql2/result_spec.rb]: Fair, fair.

**Comment by sodabrew** [comment]
And we're off to the races!

**Comment by tamird** [comment]
:confetti_ball:

**Pull request merged**